src/typechecker/FStar.TypeChecker.Util.fs

```diff
@@ -1011,26 +1011,96 @@ let check_trivial_precondition env c =
   
   ct, vc, Env.guard_of_guard_formula <| NonTrivial vc
 
-let maybe_coerce_bool_to_type env (e:term) (lc:lcomp) (t:term) : term * lcomp =
-    if env.is_pattern then e, lc else
+let coerce_with (env:Env.env) (e : term) (lc : lcomp) (f : lident) (ty : typ) : term * lcomp =
+    match Env.try_lookup_lid env f with
+    | Some _ ->
+        if Env.debug env (Options.Other "Coercions") then
+            BU.print1 "Coercing with %s!\n" (Ident.string_of_lid f);
+        let coertion = S.fvar (Ident.set_lid_range f e.pos) (Delta_constant_at_level 1) None in
```

Comment:
This leaves the universe instantiation for `hide` and `reveal` to the second phase (right?). Whereas in all other code universe instantiations happen in the first phase. It would be good to keep it that way, and instantiate the universes here (e.g. caller can pass a list of universes to instantiate the `fvar` with).

Reply:
Ah, true, and also the implicits, right? I fixed that.